sync/src/main/java/tech/pegasys/teku/sync/forward/multipeer/SyncTargetSelector.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2021 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.sync.forward.multipeer;
+
+import java.util.Optional;
+import java.util.function.Predicate;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.datastructures.blocks.SignedBeaconBlock;
+import tech.pegasys.teku.datastructures.blocks.SlotAndBlockRoot;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.statetransition.util.PendingPool;
+import tech.pegasys.teku.storage.client.RecentChainData;
+import tech.pegasys.teku.sync.forward.multipeer.chains.TargetChain;
+import tech.pegasys.teku.sync.forward.multipeer.chains.TargetChains;
+
+public class SyncTargetSelector {
+  private static final Logger LOG = LogManager.getLogger();
+
+  static final int MIN_PENDING_BLOCKS = 5;
+  private final RecentChainData recentChainData;
+  private final PendingPool<SignedBeaconBlock> pendingBlocks;
+  private final TargetChains finalizedChains;
+  private final TargetChains nonfinalizedChains;
+  private final int syncThresholdInSlots;
+
+  public SyncTargetSelector(
+      final RecentChainData recentChainData,
+      final PendingPool<SignedBeaconBlock> pendingBlocks,
+      final TargetChains finalizedChains,
+      final TargetChains nonfinalizedChains,
+      final int syncThresholdInSlots) {
+    this.recentChainData = recentChainData;
+    this.pendingBlocks = pendingBlocks;
+    this.finalizedChains = finalizedChains;
+    this.nonfinalizedChains = nonfinalizedChains;
+    this.syncThresholdInSlots = syncThresholdInSlots;
+  }
+
+  public Optional<SyncTarget> selectSyncTarget(
+      final Optional<SyncTarget> activeSyncTarget, final boolean inSyncMode) {
```

Comment:
Maybe we should just pass the `Optional<InProgressSync>` into this method and derive everything from that.

Reply:
`inSyncMode` is intended to flag if the beacon node is currently marked as syncing (ie printing `Sync Event` instead of `Slot Event`).   Turns out we can already derive this from `activeSyncTarget` since it's only non-empty if the sync is still active so this boils down to just whether the active sync is speculative or not which we already know.